### PR TITLE
[HttpFoundation] Add support for mysql unix_socket and charset in PdoSessionHandler::buildDsnFromUrl

### LINKED ISSUE #40290: PdoSessionHandler::buildDsnFromUrl not compatible with Doctrine Connection URL
**Symfony version(s) affected**: All Versions
**Description**
The Connection URL support in `PdoSessionHandler` does not support the `unix_socket` connection options like the doctrine connection factory does.

**How to reproduce**

`.env`

````
DATABASE_URL=mysql://myusername:mypassword@localhost/mydb?unix_socket=/run/proxysql/proxysql.sock
````

`config/packages/doctrine.yaml`
```yaml
doctrine:
  dbal:
    driver:   "pdo_mysql"
    url: '%env(resolve:DATABASE_URL)%'
```

`config/services.yaml`
```yaml
  Symfony\Component\HttpFoundation\Session\Storage\Handler\PdoSessionHandler:
    arguments:
      - '%env(DATABASE_URL)%'
      - { db_table: 'Session' }
```

The doctrine connection and components direly using the PDO Connection will work. Setup of the Session Handler will fail because the unix_socket part will be discarded.

**Possible Solution**
Add support for `unix_socket` in [PdoSessionHandler::buildDsnFromUrl](
https://github.com/symfony/http-foundation/blob/933b4cba7ab9abdb2472c406c433f3f78e3f9bbd/Session/Storage/Handler/PdoSessionHandler.php#L492)

Would you accept a PR adding `unix_socket` support to `buildDsnFromUrl`?

### PR DESCRIPTION
**Pull request by bcremer**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes 
| Deprecations? | no 
| Tickets       | Fix https://github.com/symfony/symfony/issues/40290 
| License       | MIT
| Doc PR        | symfony/symfony-docs#...

### CODE REVIEW

**Comment by bcremer** [comment]
Travis CI fail seems to be unrelated.

File: src/Symfony/Component/HttpFoundation/Tests/Session/Storage/Handler/PdoSessionHandlerTest.php [hunk 1]
```diff
@@ -332,6 +332,7 @@ public function testUrlDsn($url, $expectedDsn, $expectedUser = null, $expectedPa
     public function provideUrlDsnPairs()
     {
         yield ['mysql://localhost/test', 'mysql:host=localhost;dbname=test;'];
+        yield ['mysql://localhost/test?unix_socket=socket.sock&charset=utf8mb4', 'mysql:host=localhost;dbname=test;unix_socket=socket.sock;charset=utf8mb4;'];
```
Comment (Nyholm) [inline src/Symfony/Component/HttpFoundation/Tests/Session/Storage/Handler/PdoSessionHandlerTest.php]: `unix_socket` cannot be used with `host` or `port`. See https://www.php.net/manual/en/ref.pdo-mysql.connection.php

```suggestion
        yield ['mysql://localhost/test?unix_socket=socket.sock&charset=utf8mb4', 'mysql:dbname=test;unix_socket=socket.sock;charset=utf8mb4;'];
```

**Review by Nyholm — CHANGES_REQUESTED** [review]
Thank you. 

Just a small thing and I'll be happy to merge.

**Review by derrabus — APPROVED** [review]
(after @Nyholm's comment has been addressed)

**Comment by bcremer** [comment]
Added a commit but I'm not sure if the added complexity is worth it. 

The documentation states:
>    The MySQL Unix socket (shouldn't be used with host or port).

Still it works fine when port and host are given. So I would argue it's up the the user not specify a host that's not localhost when using the unix_socket param.

**Comment by Nyholm** [comment]
I moved around the logic in a way so it became more clear and (hopefully) less complex. Basically: "If you are doing `unix_socket`, do the complex code. If not, just continue as normally".

**Review by Nyholm — APPROVED** [review]
@bcremer I would be happy to see what you think.

**Comment by bcremer** [comment]
@Nyholm Looks like an reasonable improvement for me.

Maybe add the following testcase to be sure the charset option works also when unix_socket is not given:

``` 
yield ['mysql://localhost/test?charset=utf8mb4', 'mysql:charset=utf8mb4;dbname=test'];
```

**Comment by Nyholm** [comment]
Great idea. 
Yes, please do

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php [hunk 2]
```diff
@@ -486,10 +486,32 @@ private function buildDsnFromUrl(string $dsnOrUrl): string
             $driver = substr($driver, 4);
         }
 
+        $dsn = null;
```
Comment (derrabus) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php:489]: We don't need this. 🤓

Comment (Nyholm) [reply src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php:489]: We do.. Or we will reference an undefined variable. 

```php
case 'pgsql':
  $dsn = $dsn ?? 'pgsql:';
```

Is it okey to do `??` on a variable not defined?

**Comment by nicolas-grekas** [comment]
Thank you @bcremer.

**Pull request merged**